lexibank_sagartst.py

```diff
@@ -24,78 +24,60 @@ class OurLanguage(Language):
             args.writer.add_concept(
-                ID=c.concepticon_id,
-                TBL_ID=c.attributes['huang_1992_1820'],
-                Name=c.english,
-                Coverage=c.attributes['coverage'],
-                Concepticon_ID=c.concepticon_id
+                ID=idx,
+                TBL_ID=concept.attributes['huang_1992_1820'],
+                Name=concept.english,
+                Coverage=concept.attributes['coverage'],
+                Concepticon_ID=concept.concepticon_id,
+                Concepticon_Gloss=concept.concepticon_gloss
             )
-            concept2id[c.english] = c.concepticon_id
-
-        source_dict, langs_dict = {}, {}
-        for l in self.languages:
-            source_dict[l['Name']] = l['Source']
-            langs_dict[l['Name']] = l['ID']
-
-        wl.output(
-            'tsv',
-            filename=str(self.raw_dir / 'sino-tibetan-cleaned'),
-            subset=True,
-            rows={"ID": "not in "+str(data['blacklist'])}
-        )
-
-        for k in progressbar(wl, desc='wl-to-cldf'):
-            if wl[k, 'tokens']:
+            concepts[concept.english] = idx
+        languages, sources = {}, {}
+        for language in self.languages:
+            args.writer.add_language(**language)
+            languages[language['Name_in_Source']] = language['ID']
+            sources[language['Name_in_Source']] = language['Source']
+        for idx in progressbar(wl, desc='cldfify'):
+            if wl[idx, 'tokens']:
                 row = args.writer.add_form_with_segments(
-                    Language_ID=langs_dict.get(
-                        data['taxa'].get(wl[k, 'doculect'], wl[k, 'doculect'])),
-                    Parameter_ID=concept2id[wl[k, 'concept']],
-                    Value=wl[k, 'entry_in_source'].strip()
-                        or ''.join(wl[k, 'tokens']) or wl[k, 'ipa'],
-                    Form=wl[k, 'ipa'] or wl[k, 'entry_in_source'] or ''.join(wl[k, 'tokens']),
-                    Segments=wl[k, 'tokens'],
-                    Source=source_dict.get(data['taxa'].get(
-                        wl[k, 'doculect'], wl[k, 'doculect'])).split(','),
-                    Comment=wl[k, 'note'],
-                    Cognacy=wl[k, 'cogid'],
-                    Loan=True if wl[k,'borrowing'].strip() else False
+                    Language_ID=languages[wl[idx, 'doculect']],
+                    Parameter_ID=concepts[wl[idx, 'concept']],
+                    Value=wl[idx, 'entry_in_source'].strip()
+                        or ''.join(wl[idx, 'tokens']) or wl[idx, 'ipa'],
+                    Form=wl[idx, 'ipa'] or wl[idx, 'entry_in_source'] or ''.join(wl[idx, 'tokens']),
+                    Segments=' '.join(
+                        [segments.get(
+                            x,
+                            segments.get(x.split('/')[1] if '/' in x else x, x)
+                            ) for x in wl[idx, 'tokens']]).split(),
```

Comment:
I agree with @tresoldi that we should formalize this process. At least, I'd say segment corrections should be read from `etc/segments.csv` - not a JSON file for this type of data and CSV for all others.
But probably the whole replacement process should happen in `add_form_with_segments`. I guess we wouldn't want to correct segments that have been produced from an orthography profile, right?

Reply:
Yes, it is only for one case: people work in segmented form, e.g., in edictor, but don't pay attention to CLTS. The problem is that we do not yet know:

1. how many datasets there will be in this form
2. how many different ways of treating this problem will be there

So my idea would be: let's not rush with formalizing this, but rather collect use-case examples. I can switch to csv/tsv here easily. We keep the naming practice `segments`, and we can later, when we realize we have some 10 datasets in this form, get back and make another pylexibank formspec that would account for this, if we realize we CAN modularize these corrections.

Reply:
I've seen `dict`s for such segment mappings in several datasets, so that seems to happen - and be solved in the same way - more often.

I don't think `FormSpec` can do anything here, though. It's more like `lexemes.csv`, i.e. wholesale replacement, which happens in the corresponding `LexibankWriter` methods. (See my proposal https://github.com/lexibank/pylexibank/pull/161)

Reply:
Indeed, one alternative would be to have the corrected segments in a `SEGMENTS` column in `lexemes.csv`. At least in those I was involved, the offending entries were at most 60 or 70.

The rationale would be that, if the segments are so bad that pretty much every single lexeme would need a hard-coded correction, it would be better to just disregard the source segmentation and do our own profile. I believe that a list of segment replacements is more transparent, but would be fine with this different approach.

Reply:
@tresoldi, the problem is that lexemes is invoked much earlier in the process. So this is not an option here. The, we could just resegment from the beginning.

Reply:
Ok, so the analogous process to `lexemes.csv` would replacing **full** segmentations, not just sub-lists with the segmentation. While I think this would be more transparent, there may be the orthogonal use case of many segmentations sharing just one "bad" segment - in which case a `dict` of one-character mappings would be better suited.

So here's a potential compromise: Interpret `segments.csv` as replacements of sub-lists in segmentations as described [here](https://github.com/lexibank/pylexibank/pull/161#discussion_r352100260), but add "review instructions" to our best practice guide, along the lines of
- "no more than 10 simple replacements in segments.csv"
- "avoid small-sublist replacements" - i.e. prefer either single token, or full-segmentation